pyomo/contrib/mindtpy/util.py

```diff
@@ -288,6 +294,66 @@ def generate_norm_inf_objective_function(model, setpoint_model, discrete_only=Fa
+            var.unfix()
+    main_objective = next(
+        temp_model.component_data_objects(Objective, descend_into=False))
+    main_objective.activate()
+    temp_model.MindtPy_utils.deactivate()
+    TransformationFactory('core.relax_integer_vars').apply_to(temp_model)
+    # TODO: PyNumero does not support discrete variables
+    # So PyomoNLP should operate on setpoint_model
+
+    # Implementation 1
+    nlp = PyomoNLP(temp_model)
+    obj_grad = nlp.evaluate_grad_objective().reshape(-1, 1)
+    jac = nlp.evaluate_jacobian().toarray()
+    hess_lag = nlp.evaluate_hessian_lag().toarray()
+    dual_values = np.array(list(
+        temp_model.dual[c] for c in nlp.get_pyomo_constraints())).reshape(-1, 1)
+    jac_lag = obj_grad + jac.transpose().dot(dual_values)
+    first_order_term = sum(float(jac_lag[nlp.get_primal_indices([temp_var])[0]]) * (var - var.value) for var,
+                           temp_var in zip(model.MindtPy_utils.variable_list[:-1], temp_model.MindtPy_utils.variable_list[:-1]))
+    second_order_term = 0.5 * sum((var_i - var_i.value) * float(hess_lag[nlp.get_primal_indices([temp_var_i])[0]][nlp.get_primal_indices([temp_var_j])[0]]) * (var_j - var_j.value)
+                                  for var_i, temp_var_i in zip(model.MindtPy_utils.variable_list[:-1], temp_model.MindtPy_utils.variable_list[:-1])
+                                  for var_j, temp_var_j in zip(model.MindtPy_utils.variable_list[:-1], temp_model.MindtPy_utils.variable_list[:-1]))
+
+    # Implementation 2
+    # nlp = PyomoNLP(temp_model)
+    # obj_grad = nlp.extract_subvector_grad_objective(
+    #     temp_model.MindtPy_utils.variable_list[:-1])
+    # jac = nlp.extract_submatrix_jacobian(temp_model.MindtPy_utils.variable_list[:-1],
+    #                                      temp_model.MindtPy_utils.constraint_list[:-1]).toarray()
+    # hess_lag = nlp.extract_submatrix_hessian_lag(temp_model.MindtPy_utils.variable_list[:-1],
+    #                                              temp_model.MindtPy_utils.variable_list[:-1]).toarray()
+    # dual_values = np.array(list(
+    #     temp_model.dual[c] for c in temp_model.MindtPy_utils.constraint_list[:-1]))
+    # jac_lag = obj_grad + jac.transpose().dot(dual_values)
+    # first_order_term = sum(jac_lag[i] * (var - var.value)
+    #                        for i, var in enumerate(model.MindtPy_utils.variable_list[:-1]))
+    # second_order_term = 0.5 * sum((var_i - var_i.value) * float(hess_lag[i][j]) * (var_j - var_j.value)
+    #                               for i, var_i in enumerate(model.MindtPy_utils.variable_list[:-1])
+    #                               for j, var_j in enumerate(model.MindtPy_utils.variable_list[:-1]))
+
+    if config.add_regularization == "grad_lag":
+        return Objective(expr=first_order_term, sense=minimize)
+    elif config.add_regularization == "hess_lag":
+        return Objective(expr=first_order_term + second_order_term, sense=minimize)
```

Comment:
I would only construct the second-order term (calling the hessian of the lagrangian and writing the objective in memory) if this condition is satisfied. In the linear case, we do not want to go through all this trouble.

Reply:
Do you mean we only use Pynumero to calculate the second term? For the first term, we still use the jacobian in MindtPy?

Reply:
No, I meant exactly what you did in the code.
`elif config.add_regularization == "hess_lag":`  Calculate the second-order things